### DISCUSSION TITLE
Control timeout behavior to enable offline mode?

### DISCUSSION THREAD

**Discussion by 0x079** [redis/rueidis#192] [Ideas]
Please correct me if I'm wrong. Currently looks like if the client loses connection to the Redis database, and an uncached key is requested, Rueidis client will just hang there for a long time. Is there a way to control this behavior? For example, return nil immediately when that happens or with a timeout that can be configured?

This is very useful for enabling offline mode so the client will still be able to operate in the event of a redis outage. Otherwise we'll have to bring our own caching layer in front of Rueidis and that leads to a lot of complexity and duplicate caching.

**Comment by rueian** [comment]
Hi @0x079,

I believe it is the underlying retry mechanism make it looks like hanging. There are two ways to adjust the retry behavior: 

1. Setting timeout on request ctx, or canceling it directly. See https://github.com/rueian/rueidis#context-cancellation
2. Disable the retry mechanism here: https://github.com/rueian/rueidis/blob/df49264ce6e83da6573842ee2ff50974a45d2bfb/rueidis.go#L125

**Reply by 0x079** [reply]
Thank you for the response.
It looks like once it loses connection to the DB, it'll just return `connection refused. Missing key:...` error even if the key is already cached.
Setting `DisableRetry` to true will make it return an error instead of hanging. Is there a way to get it respond with what's in its cache when the DB connection is down?

**Reply by rueian** [reply]
Currently, no. According to https://redis.io/docs/manual/client-side-caching, once the connection is lost, the cache should be clear to avoid getting stale data.

If you really want to serve stale data, I am working on allowing user to bring their own cache implementation in the future. Please stay tuned.

**Reply by rueian** [reply]
Hi @0x079,

Now you can specify `NewCacheStoreFn` in `rueidis.NewClient` to bring your own cache implementation.
Please checkout https://github.com/redis/rueidis/blob/main/cache_test.go for the implementation detail.
